lib/src/metta/runner/modules/mod_names.rs

```diff
@@ -104,16 +193,38 @@ impl ModNameNode {
         }
     }
 
+    /// Updates the ModId of an existing node, or creates the node if it doesn't exist
+    /// Does NOT recursively add multiple nodes
+    pub fn update(&mut self, name: &str, mod_id: ModId) -> Result<(), String> {
+        self.update_in_layered(&mut[], name, mod_id)
+    }
+
+    /// Layered version of [Self::update].  See [Self::resolve_layered] for details
+    pub fn update_in_layered(&mut self, subtrees: &mut[(&str, &mut Self)], name: &str, mod_id: ModId) -> Result<(), String> {
+        if name == TOP_MOD_NAME || name.len() == 0 {
+            self.mod_id = mod_id;
+            return Ok(());
+        }
+        let (parent_node, mod_name) = self.parse_parent_layered_mut(subtrees, name)?;
+        if mod_name == TOP_MOD_NAME || mod_name == SELF_MOD_NAME || mod_name.len() == 0 {
+            return Err(format!("illegal module name: {name}"));
+        }
+        match parent_node.get_child_mut(mod_name) {
+            Some(node) => node.mod_id = mod_id,
+            None => parent_node.add_child_node(mod_name.to_string(), Self::new(mod_id))
+        }
+        Ok(())
+    }
+
     /// Adds a single new node to the tree.  Does NOT recursively add multiple nodes
     ///
-    /// Reuturns `true` if the node was sucessfully added, otherwise `false`.  If an entry
-    /// already exists at that name, the existing entry will be replaced
+    /// If an entry already exists at that name, the existing entry will be replaced
     pub fn add(&mut self, name: &str, mod_id: ModId) -> Result<(), String> {
         self.merge_subtree_into(name, Self::new(mod_id))
     }
 
-    /// Adds a single new node to a layered tree.  See [Self::add] and [Self::resolve_layered] for
-    /// details about behavior
+    /// Adds a single new node to a layered tree.  See [Self::resolve_layered] for details
+    #[allow(dead_code)] //NOTE: This function "feels" like it belongs in a complete API, and might be used in the future.  See discussion on [ModuleInitFrame::add_module_to_name_tree]
```

Comment:
> Am I right this PR introduces _layered API for a future use and...

Actually the `_layered` API has been there since the beginning.  This PR actually uses it less heavily than the existing code, on the account of requiring each module to load sub-modules within its own scope.  See https://github.com/luketpeterson/hyperon-experimental/blob/4d0a92fd6032616a218713246b4f64520365f606/lib/src/metta/runner/modules/mod.rs#L573

> this is the reason why all _layered methods are public?

They're public within the crate.  But ModNameNode itself is never exposed to the users of Hyperon.

> ...introduce `ModNameNode` trait and have two implementations... If it makes sense from you perspective it is better to make in a separate PR I believe as change will be massive

I like your idea because it makes the interface cleaner.  It probably wouldn't be too bad in terms of refactoring.  But I also don't think it's a high priority because this is all internal code.

Reply:
Ok, I am not sure how non-trivial this refactoring should be and if it even possible.